### ISSUE TITLE
Layernorm runs out of L1 for interleaved inputs (using only one core)

### ISSUE THREAD

**Issue by johanna-rock-tt** [tenstorrent/tt-metal#6732]
**Describe the bug**
When using layernorm with interleaved input, it runs out of L1 memory. The reason for this is that it will only select one core per number-of-tiles-in-row-dim of the input for the interleaved version. Thus if the input width is too big to fit on one core it will run out of L1. This is the case for common dims such as Falcon40b, where the input is of size [1, 1, S, 8192].

**To Reproduce**

1. Checkout `jrock/prefill-functional@1ddac50e45fdcb2595407ce42a345283a26c3586`
2. Run unit test `pytest models/experimental/falcon40b/tests/ops/test_falcon_layernorm.py::test_FalconLayernorm_inference`

**Additional Context**
We can use sharded inputs as a workaround for now, but we should have an interleaved functional version of layernorm for common dims as in Falcon40b and Llama2.

**Comment by johanna-rock-tt** [comment]
FYI: @jliangTT @miacim @cglagovich @xuncaiTT @kevinmiTT11

**Comment by yugaoTT** [comment]
for this dim, we might be able to try inplace input tensor and output tensor. However, it will only run in a single core. The way LN interleaved is designed, is each core takes a full row and cannot split it multiple cores

**Comment by johanna-rock-tt** [comment]
What's your estimate for extending this interleaved implementation to use multiple cores per row? Is that a small-ish or big change? This will be a common case for larger LLMs and running out of memory because we only use 1 out of 64 available cores doesn't seem like a good solution in the general case.

**Comment by yugaoTT** [comment]
Just to be clear, the interleaved version is not only using one core, if you have more rows or batches it would be using multiple cores. It would be big change though, to support splitting rows across cores

**Comment by yugaoTT** [comment]
@jvasilje fyi

**Comment by johanna-rock-tt** [comment]
Yeah it's an issue for large models (large hidden dim) but independent of the seq length.

**Comment by jliangTT** [comment]
@yugaoTT , do you mind letting me know what this is blocked on? thanks in advanced.

**Comment by yugaoTT** [comment]
@jliangTT I thought we haven't got a conclusion on whether this feature (mcast support, split rows to multiple cores) in interleaved LayerNorm is still needed? Last time I remember (perhaps I remembered wrong), since we have shared version working for this, interleaved is not necessary.

**Comment by jliangTT** [comment]
@johanna-rock-tt , what is the latest on the figuring out whether the sharded version of the layernorm can be used. I also don't have memory on where we are at with that too.

**Comment by johanna-rock-tt** [comment]
Currently for falcon40b and 2k sequence length we need to partially slice the activations and convert from interleaved DRAM to sharded L1; we need 4 slices to fit into L1. In total this takes 0.52 ms for the parial interlevead to sharded and back to interleaved ops (excluding the actual layernorm which takes 0.27 ms for the 4 slices added up).

Performance figures:
- 2x layernorm per decoder layer, together they take ~1.55 ms (1 ms load+convert data, 0.55 ms layernorm)
- The layernorms take ~7.4 % of model execution
- Our per layer target is ~2.6 ms (this target is very ambitious, but our current layernorm takes ~60% of the target)

I believe at the moment we still have bigger bottlenecks than the layernorm, but there seems to be some optimization potential nonetheless. I'd suggest to down-prioritize this issue for now but keep it in the backlog and revisit at a later point in time.

**Comment by yugaoTT** [comment]
@johanna-rock-tt I think with the distributed LN, this issue is no longer needed anymore?

**Comment by johanna-rock-tt** [comment]
Yes, I'll close it.
